### DISCUSSION TITLE
I2S endianness clarification, do we need to convert byte order?

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#9278] [Libraries & Drivers] [answered]
The `I2S.readinto(buf)` and `I2S.write(buf)` methods decribed in the [official Micropython I2S docs](https://docs.micropython.org/en/latest/library/machine.I2S.html?highlight=i2s#machine.I2S) state that the 'buf' object byte ordering is little-endian.
Do we need to take care of this because as I remember the I2S standard sends big endian words over the I2S bus. Does this mean that when reading from the DMA buffer we receive big endians? Do we need to convert to little endian to store it in a buffer object. (similar for writing buffer objects (little endian), do we need to convert them to big endian prior to writing to the DMA buffer?)

**Comment by miketeachman** [comment, marked as answer]
No byte order conversion is needed.   For the `read(buf)` method, `buf` will contain samples in little endian format.  For the `write(buf)` method, sample data in `buf` must be arranged in little endian format.

Here is an example for sample data received from an I2S microphone using the `read(buf)` method.
```
I2S SD line                                                                   MicroPython buf[]
=======                                                                       ============
1st byte (Most significant byte, first to be transmitted from the mic)        buf[i+3]
2nd byte                                                                      buf[i+2]
3rd byte (Least significant byte of data transmitted from the mic)            buf[i+1]
4th byte (mic is in tri-state for this data, needs the pull-down to read 0)   buf[i]              
```

**Reply by saraverbeecke** [reply]
@miketeachman Ok crystal clear, thanks for your answer.
